mlflow/models/utils.py

```diff
@@ -54,61 +56,78 @@ def __init__(self, input_example: ModelInputExample):
-                    raise TensorsNotSupportedException("Row '{0}' has shape {1}".format(i, x.shape))
-            if all([_is_scalar(x) for x in input_example]):
-                input_example = pd.DataFrame([input_example], columns=range(len(input_example)))
-            else:
-                input_example = pd.DataFrame(input_example)
-        elif isinstance(input_example, np.ndarray):
-            if len(input_example.shape) > 2:
-                raise TensorsNotSupportedException(
-                    "Input array has shape {}".format(input_example.shape)
-                )
-            input_example = pd.DataFrame(input_example)
-        elif not isinstance(input_example, pd.DataFrame):
-            try:
-                import pyspark.sql.dataframe
-
-                if isinstance(input_example, pyspark.sql.dataframe.DataFrame):
-                    raise MlflowException(
-                        "Examples can not be provided as Spark Dataframe. "
-                        "Please make sure your example is of a small size and "
-                        "turn it into a pandas DataFrame by calling toPandas "
-                        "method."
+                return {"instances": input_example.tolist()}
+
+        def _handle_dataframe_input(input_ex):
+            if isinstance(input_ex, dict):
+                if all([_is_scalar(x) for x in input_ex.values()]):
+                    input_ex = pd.DataFrame([input_ex])
+                else:
+                    raise TypeError(
+                        "Data in the dictionary must be scalar or of type numpy.ndarray"
                     )
-            except ImportError:
-                pass
-            raise TypeError(
-                "Unexpected type of input_example. Expected one of "
-                "(pandas.DataFrame, numpy.ndarray, dict, list), "
-                "got {}".format(type(input_example))
-            )
+            elif isinstance(input_ex, list):
+                for i, x in enumerate(input_ex):
+                    if isinstance(x, np.ndarray) and len(x.shape) > 1:
+                        raise TensorsNotSupportedException(
+                            "Row '{0}' has shape {1}".format(i, x.shape)
+                        )
+                if all([_is_scalar(x) for x in input_ex]):
+                    input_ex = pd.DataFrame([input_ex], columns=range(len(input_ex)))
+                else:
+                    input_ex = pd.DataFrame(input_ex)
+            elif not isinstance(input_ex, pd.DataFrame):
+                try:
+                    import pyspark.sql.dataframe
+
+                    if isinstance(input_example, pyspark.sql.dataframe.DataFrame):
+                        raise MlflowException(
+                            "Examples can not be provided as Spark Dataframe. "
+                            "Please make sure your example is of a small size and "
+                            "turn it into a pandas DataFrame or a tensor."
```

Comment:
nit: I think we can drop `or a tensor` from the message, spark dataframes naturally map to pandas DataFrames and tensor may be confusing.